### ISSUE TITLE
Tree-sitter folding

### ISSUE THREAD

**Issue by 314eter** [atom/atom#17596]
### Prerequisites

* [x] Reproduced the problem in Safe Mode: https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
* [x] Followed all applicable steps in the debugging guide: https://flight-manual.atom.io/hacking-atom/sections/debugging/
* [x] Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
* [x] Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
* [x] Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

When a tree-sitter grammar contains a foldable construct with more than one line before the folding point, there are two folding arrows displayed. The first one one the line where the construct begins, and the second one on the line with the folding point. When folded, the first one does nothing.

### Steps to Reproduce

1. Enable tree-sitter parsers and install language-python.
2. Open a python file with
```python
if True or \
   False:
    pass
```
3. Click one of the folding arrows
![unfolded](https://user-images.githubusercontent.com/890465/42032224-28d2777c-7ad9-11e8-9d0b-0a1986d8a08d.png)
![folded](https://user-images.githubusercontent.com/890465/42032234-2d9a7c46-7ad9-11e8-8dcd-825499ff1cdd.png)

**Expected behavior:** There should be only one folding arrow for each possible fold

**Actual behavior:** There are two arrows, and the first one does not work as expected

### Additional Information

The example is in python, but this is not a problem with language-python, and there probably are examples in other languages too.

### Versions

    Atom    : 1.28.0
    Electron: 2.0.3
    Chrome  : 61.0.3163.100
    Node    : 8.9.3

**Comment by 314eter** [comment]
I think there are two problems. The first one is that `isFoldableCache` in [tree-sitter-language-mode.js](https://github.com/atom/atom/blob/d82c3979aca051acb86ee4fe6d0d339711c4acb7/src/tree-sitter-language-mode.js#L182) gets out of sync with reality, which causes too many or too few folding arrows.

The second one is that the folding arrow is displayed on the first line of the construct when unfolded (when `isFoldableCache is correct), and on the line with the folding point when folded.

**Comment by thomasjo** [comment]
@maxbrunsfeld Are we tracking this somewhere else already? You mentioned yesterday that you want to work on code folding with Tree-sitter parsers.

**Comment by maxbrunsfeld** [comment]
We're working on making folding work with nested languages like JS inside of HTML, but are not already working on this particular bug. Thanks for the report @314eter.

**Comment by 314eter** [comment]
It seems to be fixed with
```javascript
  getFoldableRangeContainingPoint (point, tabLength, existenceOnly = false) {
    this.ensureParseTree()
    let node = this.tree.rootNode.descendantForPosition(this.buffer.clipPosition(point))
    while (node) {
      if (node.endPosition.row > point.row) {
        const range = this.getFoldableRangeForNode(node, false)
        if (range && (!existenceOnly || range.start.row == point.row)) return range
      }
      node = node.parent
    }
  }
```
but I'm not sure this solves the problem with `isFoldableCache`.

**Comment by 314eter** [comment]
@maxbrunsfeld It's not fixed by #17717.

Opening a file with
```python
if True or \
   False:
    pass
```
gives `isFoldableCache = [true, false, false]`, which is right. But then deleting the backslash-newline results in `[false, false]` instead of `[true, false]`.

Opening a file with
```python
if True or False:
    pass
```
gives `[true, false]`, but typing a backslash-newline before `False` results in `[true, true, false]`.

**Comment by 314eter** [comment]
This issue can still be reproduced.
